### PR TITLE
WFS & WCS as module

### PR DESCRIPTION
**Pull request by dmarteau**
Move WFS and WCS services as server modules.

This is the next step after #3866

### CODE REVIEW

File: src/server/services/wfs/qgswfsutils.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * WMS implementation
+ */
+
+//! WMS implementation
+namespace QgsWfs
+{
+
+  /** Return the highest version supported by this implementation
+   */
+  QString implementationVersion();
```
Comment (nyalldawson) [inline src/server/services/wfs/qgswfsutils.h]: const

**Review by nyalldawson — COMMENTED** [review]
Nice!

Don't forget to add these new folders to the DOXYGEN_INPUT folders list in docs/CMakeLists.txt so that the doxygen is generated for them

**Comment by dmarteau** [comment]
@nyalldawson  should I add the whole folders are only add the headers of interfaces exposed to python (which seems to be the case actually) ?

File: src/server/services/wcs/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  ${CMAKE_BINARY_DIR}/src/server
+  ${CMAKE_CURRENT_BINARY_DIR}
+  ../../../core 
+  ../../../core/dxf
+  ../../../core/geometry 
+  ../../../core/raster
+  ../../../core/symbology-ng
+  ../../../core/composer
+  ../../../core/layertree
+  ../../../gui
```
Comment (dmarteau) [inline src/server/services/wcs/CMakeLists.txt]: Me too, but historically there was some dependencies. Check needed.

File: src/server/services/wcs/qgswcsdescribecoverage.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    coveDescElement.setAttribute( QStringLiteral( "xmlns:xsi" ), QStringLiteral( "http://www.w3.org/2001/XMLSchema-instance" ) );
+    coveDescElement.setAttribute( QStringLiteral( "xsi:schemaLocation" ), WCS_NAMESPACE + " http://schemas.opengis.net/wcs/1.0.0/describeCoverage.xsd" );
+    coveDescElement.setAttribute( QStringLiteral( "xmlns:gml" ), GML_NAMESPACE );
+    coveDescElement.setAttribute( QStringLiteral( "xmlns:xlink" ), QStringLiteral( "http://www.w3.org/1999/xlink" ) );
+    coveDescElement.setAttribute( QStringLiteral( "version" ), QStringLiteral( "1.0.0" ) );
+    coveDescElement.setAttribute( QStringLiteral( "updateSequence" ), QStringLiteral( "0" ) );
+    doc.appendChild( coveDescElement );
+
+    //defining coverage name
+    QString coveName = QLatin1String( "" );
```
Comment (dmarteau) [inline src/server/services/wcs/qgswcsdescribecoverage.cpp]: Arf, I thought I have killed them all....

File: src/server/services/wfs/qgswfsutils.h (hunk 1 above)
Comment (dmarteau) [inline src/server/services/wfs/qgswfsutils.h]: Where do you want set a const signature ?

Comment (elpaso) [inline src/server/services/wfs/qgswfsutils.h]: @dmarteau I think Nyall was referring to implementationVersion(); unless it changes the status of the instance, it should be const.

Comment (dmarteau) [inline src/server/services/wfs/qgswfsutils.h]: @elpaso implementationVersion() is a function, not a method attached to a class instance: const does not apply here.

File: src/server/services/wcs/CMakeLists.txt (hunk 2 above)
Comment (dmarteau) [inline src/server/services/wcs/CMakeLists.txt]: This is only needed for wms which have a dependency on QgsEditorWidgetRegistry. Removed for wcs and wfs.

Comment (nyalldawson) [inline src/server/services/wcs/CMakeLists.txt]: WMS shouldn't require that either now after https://github.com/qgis/QGIS/pull/3836

File: src/server/services/wfs/qgswfsutils.h (hunk 1 above)
Comment (nyalldawson) [inline src/server/services/wfs/qgswfsutils.h]: @dmarteau ah - sorry for the noise! ignore that comment!

**Comment by nyalldawson** [comment]
@dmarteau should I add the whole folders are only add the headers of interfaces exposed to python (which seems to be the case actually) ?

I'd add the whole folder. You can mark classes you don't want as public API by wrapping them with

///@cond PRIVATE


///@endcond PRIVATE

The benefit with adding the whole folders is that the doxygen test can properly check those new classes and flag any with missing doc strings.

**Comment by dmarteau** [comment]
@nyalldawson actually there is no really api in the modules but only implementation details: each module implement QgsService overrides.

**Comment by nyalldawson** [comment]
@dmarteau ok, fair enough.

**Comment by dmarteau** [comment]
@rldhont  any review ?

**Review by rldhont — APPROVED** [review]

**Pull request merged**